### PR TITLE
Fix 2455: Check auth for pictrs when instance is private.

### PR DESCRIPTION
**Pull request by sam365724**
https://github.com/LemmyNet/lemmy/issues/2455

### CODE REVIEW

**Comment by Nutomic** [comment]
You need to run `cargo +nightly fmt --all`

File: crates/api_common/src/utils.rs [hunk 1]
```diff
@@ -293,6 +293,20 @@ pub async fn check_private_instance(
...
+  let site = blocking(pool, Site::read_local_site).await?;
+    // The site might not be set up yet
+    if let Ok(site) = site {
+      if site.private_instance {
+        return Ok(true);
+      }
+    }
+    return Ok(false);
+}
+
```
Comment (Nutomic) [inline crates/api_common/src/utils.rs:309]: This helper function isnt helping much, and is only used in one place. I would simply move this logic into images.rs directly.

File: crates/routes/src/images.rs [hunk 2]
```diff
@@ -123,6 +126,18 @@ async fn full_res(
...
+  // block access to images if instance is private and unauthorized, public
+  let is_priv_inst = is_private_instance(context.pool()).await?;
+  if true == is_priv_inst {
+      let jwt = req
+      .cookie("jwt")
+      .expect("No auth header for picture access.");
+      
+      if Claims::decode(jwt.value(), &context.secret().jwt_secret).is_err() {
+        return Ok(HttpResponse::Unauthorized().finish());
+      };
```
Comment (Nutomic) [inline crates/routes/src/images.rs:138]: I would simply call `get_local_user_view_from_jwt()` here.

Comment (sam365724) [reply crates/routes/src/images.rs:141]: Sure, done :)

File: crates/routes/src/images.rs [hunk 3]
```diff
@@ -123,6 +128,19 @@ async fn full_res(
...
   context: web::Data<LemmyContext>,
 ) -> Result<HttpResponse, Error> {
+  // block access to images if instance is private and unauthorized, public
+  let site = blocking(context.pool(), Site::read_local_site).await?;
+  // The site might not be set up yet
+  if let Ok(site) = site {
+    if site.private_instance {
+      let jwt = req
+      .cookie("jwt")
+      .expect("No auth header for picture access");
```
Comment (dessalines) [inline crates/routes/src/images.rs:138]: No expect, do a `.map_err(error::ErrorUnauthorized)`

File: crates/routes/src/images.rs [hunk 4]
```diff
@@ -123,6 +128,19 @@ async fn full_res(
...
+      if get_local_user_view_from_jwt(jwt.value(), context.pool(), context.secret()).await.is_err() {
+        return Ok(HttpResponse::Unauthorized().finish());
```
Comment (dessalines) [inline crates/routes/src/images.rs:140]: I think its better to return the `Error(error::ErrorUnauthorized)`

**Review by dessalines — CHANGES_REQUESTED** [review]
This requires 1-2 DB reads for every single picture requested now, reading the site to check if its a private instance, reading from local_user to make sure they're actually a user.

That performance hit scares me, because some pages like Home can request hundreds of pictures.

I don't think this is on you to fix tho, its on us to create a in-memory singleton of local site settings, that gets updated on every save. Then we could at least reduce this to one read validating the local user, that should only get run on a private instances.

File: crates/routes/src/images.rs (hunk 3 above)
Comment (sam365724) [reply crates/routes/src/images.rs]: OK, that should be done upload() on line [92](https://github.com/LemmyNet/lemmy/blob/0bec040c4e8d6822755900689af8f2359f6a93ab/crates/routes/src/images.rs#L97) as well i think?

File: crates/routes/src/images.rs (hunk 4 above)
Comment (sam365724) [reply crates/routes/src/images.rs:140]: I think in both cases we should return a HttpResponse. I'm not sure how to do that honestly. 

Is replacing expect() with .map_err(error::ErrorUnauthorized) enough above?

Comment (dessalines) [reply crates/routes/src/images.rs:140]: I spose that's fine, althoug that actix errors are also HttpResponses.

**Comment by sam365724** [comment]
> This requires 1-2 DB reads for every single picture requested now, reading the site to check if its a private instance, and if it is, reading from local_user to make sure they're actually a user.
> 
> That performance hit scares me, because some pages like Home can request hundreds of pictures.
> 
> I don't think this is on you to fix tho, its on us to create a in-memory singleton of local site settings, that gets updated on every save. Then we could at least reduce this to one read validating the local user, that should only get run on a private instances.
> 
> Also the format check is failing, run `cargo +nightly fmt` to fix

I agree on the performance concern. I see up to 600 requests per user easily if you have some image intense posts. 

Thanks for the review, i did run the formatter now.

File: crates/routes/src/images.rs (hunk 4 above)
Comment (sam365724) [reply crates/routes/src/images.rs:140]: @dessalines I'm getting errors about types not being correct. I'm really unsure how to do that, and need help on those - sorry not yet familiar with those rust libs.

Comment (dessalines) [reply crates/routes/src/images.rs:140]: There are other examples in this file. If you run `cargo check`, it should tell you the exact error.

**Review by dessalines — APPROVED** [review]

**Pull request merged**